File: src/main/java/j/chain_of_resposibility/inheritance/Support.java
package j.chain_of_resposibility.inheritance;

import j.chain_of_resposibility.Trouble;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class Support {
    private final String name;
    private final Support next;

    public Support(@NotNull String name, @Nullable Support next) {
        this.name = name;
        this.next = next;
    }

    protected abstract Support createSupport(@NotNull String name, @Nullable Support next);

    public Support withNext(@NotNull Support next) {
        return createSupport(name, next);
    }

    public void support(Trouble trouble) {
        for (Support obj = this; true; obj = obj.next) {
            if (obj.resolve(trouble)) {
                obj.done(trouble);
                break;
            } else if (obj.next == null) {
                obj.fail(trouble);
                break;
            }
        }
    }

    @Override
    public String toString() {
        return "[" + name + "]";
    }

    protected abstract boolean resolve(Trouble trouble);

    protected void done(Trouble trouble) {
        System.out.println(trouble + " is resolved by " + this + ".");
    }

    protected void fail(Trouble trouble) {
        System.out.println(trouble + " cannot be resolved.");
    }
}

File: src/main/scala/s/chain_of_resposibility/inheritance/Support.scala
package s.chain_of_resposibility.inheritance

import s.chain_of_resposibility.Trouble

abstract class Support(
    private val name: String,
    private val next: Option[Support]
) {
  def getName: String = name

  def support(trouble: Trouble): Unit = {
    if (resolve(trouble)) done(trouble)
    else if (next.isDefined) next.get.support(trouble)
    else fail(trouble)
  }

  override def toString: String = "[" + name + "]"

  protected def resolve(trouble: Trouble): Boolean

  protected def done(trouble: Trouble): Unit = {
    System.out.println(trouble.toString + " is resolved by " + this + ".")
  }

  protected def fail(trouble: Trouble): Unit = {
    System.out.println(trouble.toString + " cannot be resolved.")
  }
}
